x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java

```diff
@@ -183,16 +187,31 @@ private void getVersionedDocument(String uid, ActionListener<VersionedDocument>
     }
 
     // Package private for testing
-    void getVersionedDocument(Authentication authentication, ActionListener<VersionedDocument> listener) {
+    void getVersionedDocument(Subject subject, ActionListener<VersionedDocument> listener) {
         tryFreezeAndCheckIndex(listener).ifPresent(frozenProfileIndex -> {
-            final SearchRequest searchRequest = client.prepareSearch(SECURITY_PROFILE_ALIAS)
-                .setQuery(
-                    QueryBuilders.boolQuery()
-                        .must(QueryBuilders.termQuery("user_profile.user.username", authentication.getUser().principal()))
-                        // TODO: this will be replaced by domain lookup and reverse lookup
-                        .must(QueryBuilders.termQuery("user_profile.user.realm.name", authentication.getSourceRealm().getName()))
-                )
-                .request();
+            final BoolQueryBuilder boolQuery = QueryBuilders.boolQuery()
+                .must(QueryBuilders.termQuery("user_profile.user.username", subject.getUser().principal()));
+            if (subject.getRealm().getDomain() == null) {
+                boolQuery.must(QueryBuilders.termQuery("user_profile.user.realm.name", subject.getRealm().getName()));
+            } else {
+                final String[] domainRealms = subject.getRealm()
+                    .getDomain()
+                    .realms()
+                    .stream()
+                    .map(RealmConfig.RealmIdentifier::getName)
```

Comment:
I thought we were going to rely on the realm name-type tuple for these checks, no?
The fact that realm name is unique in a given node yml config file is not good enough to generally qualify a unique user, as I recall.

Reply:
Yes you are right. I had a different thought (along the direction that we may want ultimately rely only on name). But even that is a thing in the future and for now we should still use both name and type. I updated accordingly.